src/osd/PrimaryLogPG.cc

```diff
@@ -4248,8 +4351,41 @@ int PrimaryLogPG::trim_object(
     ctx->delta_stats.num_object_clones--;
     if (coi.is_cache_pinned())
       ctx->delta_stats.num_objects_pinned--;
-    if (coi.has_manifest())
+    if (coi.has_manifest()) {
+      set<uint64_t> no_refs;
+
+      for (auto p : snapset.clones) {
+	hobject_t clone_oid = coid;
+	if (clone_oid.snap == p) {
+	  continue;
+	}
+	clone_oid.snap = p;
+	ObjectContextRef cobc = get_object_context(clone_oid, false, NULL);
+	if (!cobc) {
+	  close_op_ctx(ctx.release());
+	  dout(0) << __func__ << ": Can not find obc " << coid << dendl;;
+	  return -ENOENT;
+	}
+
+	// check if the references is still used
+	object_info_t& oi = cobc->obs.oi;
+	if (oi.has_manifest()) {
+	  coi.manifest.build_non_intersection_set(oi.manifest.chunk_map, no_refs, NULL);
```

Comment:
The way build_non_intersection_set is written, it will only add things that don't overlap.  So calling it inside a loop won't find things that don't intersect with all other clones.. instead, it'll find things that don't intersect with at least one other clone, which isn't very useful.

1. I think this only needs to check the before and after clones.
2. I think it needs to do the first pass, where it finds the intersection, on both of those clones, and *then* do the second pass.